homeassistant/components/matrix/__init__.py

```diff
@@ -115,284 +139,274 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
+        file_stat = await aiofiles.os.stat(image_path)
+
+        _LOGGER.debug("Uploading file from path, %s", image_path)
+        async with aiofiles.open(image_path, "r+b") as image_file:
+            response = await self._client.upload(
+                image_file,
+                content_type=mime_type,
+                filename=os.path.basename(image_path),
+                filesize=file_stat.st_size,
+            )
+        if isinstance(response, UploadError):
+            _LOGGER.error("Unable to upload image to the homeserver: %s", response)
+            return
 
-    def _send_image(self, img, target_rooms):
-        _LOGGER.debug("Uploading file from path, %s", img)
+        assert isinstance(response, UploadResponse)
+        _LOGGER.debug("Successfully uploaded image to the homeserver")
+
+        content = {
+            "body": os.path.basename(image_path),
+            "info": {
+                "size": file_stat.st_size,
+                "mimetype": mime_type,
+                "w": width,
+                "h": height,
+            },
+            "msgtype": "m.image",
+            "url": response.content_uri,
+        }
 
-        if not self.hass.config.is_allowed_path(img):
-            _LOGGER.error("Path not allowed: %s", img)
-            return
-        with open(img, "rb") as upfile:
-            imgfile = upfile.read()
-        content_type = mimetypes.guess_type(img)[0]
-        mxc = self._client.upload(imgfile, content_type)
-        for target_room in target_rooms:
-            try:
-                room = self._join_or_get_room(target_room)
-                room.send_image(mxc, img, mimetype=content_type)
-            except MatrixRequestError as ex:
-                _LOGGER.error(
-                    "Unable to deliver message to room '%s': %d, %s",
-                    target_room,
-                    ex.code,
-                    ex.content,
-                )
+        for room in target_rooms:
+            await self._client.room_send(
+                room_id=room, message_type="m.room.message", content=content
+            )
+            _LOGGER.debug("Image '%s' sent to room '%s'", image_path, room)
 
-    def _send_message(self, message, data, target_rooms):
-        """Send the message to the Matrix server."""
+    async def _send_message(
+        self, message: str, target_rooms: list[RoomID], data: dict | None
+    ) -> None:
+        """Send a message to the Matrix server."""
         for target_room in target_rooms:
-            try:
-                room = self._join_or_get_room(target_room)
-                if message is not None:
-                    _LOGGER.debug(room.send_text(message))
-            except MatrixRequestError as ex:
+            response: Response = await self._client.room_send(
+                room_id=target_room,
+                message_type="m.room.message",
+                content={"msgtype": "m.text", "body": message},
```

Comment:
I have another open PR for [adding HTML support](https://github.com/home-assistant/core/pull/69951). I'm not sure what the `nio` approach here is, but it looks like it might be adding a `"format": "org.matrix.custom.html", "formatted_body": "xxx"` to your `content` object. ([docs](https://spec.matrix.org/latest/client-server-api/#mroommessage-msgtypes))

Reply:
(Replying here rather than adding a comment chain)
I appreciate the review! I'll take a look at keeping up the HTML support in this PR as well for consistency.
I'd hope mainly to get the attention of @tinloaf as a primary reviewer, though I'm not sure what more I can really do at this point, for that.